Conversion to Typescript 🐈 #1

AmadeusSlabu opened the pull request:
I do not know what else to say..., maybe give me advice on things and ask me for help if you need it ? 

# Additions

- Added types as requested.
- All files, code commented.
- Added to **README.md**.

## Minor changes

---

- Inside **eventHandlers**:

  - Changed ~~`getAllFiles`~~ to `allFiles`.
  - Line 14, changed sort to `(a, b) => (a > b ? 1 : -1)` to avoid error of type boolean/string/numbers conflicting.

- Inside **getApplicationCommands**:

  - Added `import { Client } from 'discord.js';`.
  - Line 10 added `!` `Null` check
  - Line 15 changed ~~`applicationCommands.fetch()`~~ for `applicationCommands.fetch(applicationCommands as any)` because arguments were expected and `fetch(applicationCommands as any)` gave this error

  ```
  There was an error: TypeError: Failed to parse URL from [object Object]
  ```

- Inside **getLocalCommands** :

  - Changed ~~`(\_\_dirname, ".." "etc")`~~ to `(\_\_dirname, "../etc")`.

- Inside **getAllFiles**:

  - Added `Import path from 'path'` because _`path`_ variable could not be redeclared and when simply `require('path')` not enough for the _`path`_ variable to be reused inside the code. TS does not seem to be able to recognize it and wants it redeclared, so rather than adding a different name variable imported _`path`_.

- Inside **registerCommands**:

  - Added `Import {Client} from 'discord.js'` to use Client type.
  - Line 13 _`localCommands`_ type changed to `as any[]`.
  - Line 18 _`applicationCommands`_ defined type `as ApplicationCommandManager || GuildApplicationCommandManager`.

- Inside **handleCommands**:

  - Added `import { Client, GuildMember, Interaction } from 'discord.js';`.
  - Changed ~~`(interaction.member.id)`~~ for `((interaction.member as GuildMember).id)` because _`interaction.member`_ returns either `APIInteractionGuildMember` or `GuildMember` or `Null` the first and the last not being useable by the _`.id`_ command. Did the same for line 29 and 42.

- Inside **consoleLogs**:

  - Added `import { Client } from 'discord.js';` to define client as Client.

- Inside **get-user**:

  - Added `import { Client, CommandInteraction } from 'discord.js';`.
  - Defined _`Interaction`_ as _`CommandInteraction`_ because _`.reply`_ and _`.options`_ does not exist on type _`Interaction`_ (not specific enough for TS).
  - Line 12 defined `user` type `as any` otherwise a bunch of issues arise from user not being `String | Number | Boolean | Null | Undefined` its also why the `!` was added to remove `Null` and `Undefined` types from possibilities.

- Inside **get-standings**:

  - Added `import { Client, CommandInteraction } from 'discord.js';`.
  - Line 14 defined `user` type `as any` because of type issues again.

- Inside **book-tickets**:

  - Added `import { Client, CommandInteraction } from 'discord.js';`.
  - Line 11 defined `user` type `as any` again due to type issues, cannot assign it to type `as object` or `as string` TS doesn't recognize functions/keys/values on string and it cannot read the object if it doesn't have an inherited schema for `user` to read from.
  - Line 12 defined `location` type `as string` and added a `!` to tell TS that the variable is not null.

- Inside **announce-match**:

  - Added `import { CommandInteraction, Client } from 'discord.js';`.
  - Lines 11 and 12 defined types for `user1` and `user2` `as any` again, while adding `!` to tell TS values for them aren't Null.

- Inside **add-user**:

  - Added `import { CommandInteraction, Client } from 'discord.js';`.
  - Line 15 `user` set to `as any` and together with line 16 and 17 added `!` to tell TS the values are not gonna be `Null`.

- Inside **add-results**:

  - Added `import { CommandInteraction, Client } from 'discord.js';`.
  - Lines 17, 18 and 19 added `!` to tell TS those are not gonna be `Null` values.
  - Lines 17 and 19 defined users to `as any` type .

- Inside **add-champion**:

  - Added `import { CommandInteraction, Client } from 'discord.js';`.
  - Line 14 `user` defined type `as any` and `!` for `Null`

- Inside **source**:

  - Added `import { CommandInteraction } from 'discord.js';`

- Inside **open-ticket**:

  - Added `import { CommandInteraction } from 'discord.js';`

- Inside **eventHandlers**:

  - Changed ~~`(\_\_dirname, ".." "etc")`~~ to `(\_\_dirname, "../etc")`.

### Minor Issues

---

- Inside **eventHandlers**:

  Line 25, cannot assign function type to `eventFunction` because of the way it is constructed, code would have to be refactored so unless you want to change it, it will remain as implicit any.

- Inside **areCommandsDifferent**:

  Sorry I have no idea what's happening here so I just used TS infered types. 🤷

- Inside **getLocalCommands** :

  Line 18 const `commandObj` requires a file therefore I've let it to implicit any type.

- Inside **getAllFiles**:

  Line 7 `files` cannot be defined since it includes both directories and files and even tho it retrieves paths to said things the `.isFile` and `.isDirectory` functions are not applicable to strings.

### Comments

---

> I believe the register commands thingie and delete are not yet fully fleshed out ? Because I cannot define their type and there are certain {'etcetc'} replies when executing the commands, problem is I do not find those commands in the option of available commands.

AmadeusSlabu committed:
TS config

AmadeusSlabu committed:
TS config

AmadeusSlabu committed:
Changing a few requires to imports/exports

AmadeusSlabu committed:
Singular function exports

AmadeusSlabu committed:
areCommandsDifferent too many any

AmadeusSlabu committed:
areCommandsDifferent any++

AmadeusSlabu committed:
Installed types for discord

AmadeusSlabu committed:
Utils half-assed

AmadeusSlabu committed:
Figured out how to remove .env can't find err

AmadeusSlabu committed:
Finished changing imports and solving err

AmadeusSlabu committed:
Ya

AmadeusSlabu committed:
ya

AmadeusSlabu committed:
Types set

AmadeusSlabu committed:
nodemon change

AmadeusSlabu committed:
no

AmadeusSlabu committed:
Yea

AmadeusSlabu committed:
Refunded Imports for original

AmadeusSlabu committed:
The end

AmadeusSlabu committed:
Some comments

AmadeusSlabu committed:
Some more changes

AmadeusSlabu committed:
Only commands left to finish

AmadeusSlabu committed:
Done, remain to test and clean MD

AmadeusSlabu committed:
Brought back the original if

AmadeusSlabu committed:
Final form is complete magic

AmadeusSlabu committed:
Final push

AmadeusSlabu committed:
Added the text

AmadeusSlabu committed:
Changed eslint to hopefuly work for TS

stinkyfi commented:
Thank you for contributing, I will do my best to fully review the pull request before the end of the weekend.

AmadeusSlabu commented:
> Thank you for contributing, I will do my best to fully review the pull request before the end of the weekend.

Don't worry its my first pull request and I believe I might have to change a lot of things, like for example eslint right now shows me ( after I set it up ) that it doesn't like both imports and requires. Take your time and let me know what I can change and help you with. And have a great day ? afternoon ? something ! :3

stinkyfi commented on src/commands/info/source.ts:
```diff
@@ -0,0 +1,15 @@
+   * @param {Interaction} interaction */
+  callback: (interaction: CommandInteraction) => {
+    console.log('Get Source');
+    interaction.reply(process.env.SOURCE);
```
I get the following error when I attempt to run the bot.
```sh
There was an error: TSError: ⨯ Unable to compile TypeScript:
src/commands/info/source.ts:11:23 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(options: InteractionReplyOptions & { fetchReply: true; }): Promise<Message<boolean>>', gave the following error.
    Argument of type 'string | undefined' is not assignable to parameter of type 'InteractionReplyOptions & { fetchReply: true; }'.
      Type 'undefined' is not assignable to type 'InteractionReplyOptions & { fetchReply: true; }'.
        Type 'undefined' is not assignable to type 'InteractionReplyOptions'.
  Overload 2 of 2, '(options: string | InteractionReplyOptions | MessagePayload): Promise<InteractionResponse<boolean>>', gave the following error.
    Argument of type 'string | undefined' is not assignable to parameter of type 'string | InteractionReplyOptions | MessagePayload'.
      Type 'undefined' is not assignable to type 'string | InteractionReplyOptions | MessagePayload'.

11     interaction.reply(process.env.SOURCE);
                         ~~~~~~~~~~~~~~~~~~
```

I have this value set in my `.env` file as:

```env
SOURCE = 'https://github.com/stinkyfi/MaddenBot'
```

Do you not experience the same issue?

stinkyfi replied:
There is no error in `opent-ticket.ts`. I found that copying what that file is doing and concat an empty string solved the errror. 
```ts 
interaction.reply(process.env.SOURCE + '');
```
however during runtime of the `/source` command in discord, I get this error.

```sh
There was an error running this command: TypeError: interaction.reply is not a function
```

I will continue to debug later, just wanted to report this bug to see if you can replicate

AmadeusSlabu replied:
I will get on my pc to check rn. Also good morning, I guess.
Oh and add me on the same username on Discord if you need faster replies or want to not flood your project with comments.

AmadeusSlabu replied:
Okay step by step : 
- I did not have `SOURCE` as an example in .env and being new and all I kinda never added it in the first place.
- Second, the overloads are simply because types are `string | undefined` you need to add a `!` to the parts that have a possible undefined OR define their type `as string`.
- Thirdly yes I had a similar issue where the things inside .env were not being read hence why you are getting this error.
```
There was an error running this command: TypeError: interaction.reply is not a function
``` 
But yes I will try to replicate right now I hope I am actually useful after a coffee 🤣

stinkyfi commented on src/commands/info/open-ticket.ts:
```diff
@@ -0,0 +1,16 @@
+   * @param {Interaction} interaction */
+  callback: (interaction: CommandInteraction) => {
+    console.log('Bug/Feature Link');
+    interaction.reply(process.env.SOURCE + '/issues');
```
I get the same error at runtime as [`source.ts`](https://github.com/stinkyfi/MaddenBot/pull/11/files#r1240613450)

```sh
There was an error running this command: TypeError: interaction.reply is not a function
```

AmadeusSlabu replied:
Yes it seems to be my bad for not being capable of defining interaction type properly, it seems that client is `client` but interaction is not `interaction` type but something else, I also tried asigning it to `any` but it does not seem to solve the problem. I am sorry you have to deal with my mistakes.

stinkyfi replied:
I am going to propose removing the `/info` commands for now. I think your typescript changes are highly useful.
I can't make sense of the error right now, and worst case they can be reworked to embeded message.

actively testing the other commands

AmadeusSlabu replied:
Sorry for the dellay kind stinkyfi, I have tried some stuff now that my brain SORT of works and heres a code snippet for you to see : 
``` console.log('Get Source');
    if (interaction instanceof CommandInteraction) {
      console.log('interaction is a CommandInteraction');
    } else {
      console.log('interaction is not a CommandInteraction');
      console.log(typeof interaction);
    }
```
As for the output :
```
Connected to DB
Testbot2#5223 is online.
Get Source
interaction is not a CommandInteraction
object
There was an error running this command: TypeError: interaction.reply is not a function
``` 
The problem is not with how you wrote the code, like I have said previously the problem was either the import or the type of for interaction and as I have just showed you it seems that in this case interaction is not type of CommandInteraction, rather than discard the command all together you should keep it, I will try to find the proper type ( neither object or any work so ugh ).

stinkyfi replied:
yeah, it's a tough one. I would only suggest deleting them temporarily to get this merge request in. Creating new tickets to address them in the future. They aren't critical features but more nice to haves. Converting them to use discord.js embeded messages might be a simple enough solution as well.

AmadeusSlabu replied:
Fixed, no need to get rid of the file, the problem was interaction was running but it had nowhere to run therefore it didn't count as a function. Here is my solution :
```
import { CommandInteraction, Client } from 'discord.js';
//make sure to also import it
callback: (_client: Client, interaction: CommandInteraction)
```

stinkyfi replied:
Well done! please see my `.eslintrc.json` changes, let's get the linter working and we can merge this in.

stinkyfi commented:
I was able to get the linter to work, lots of "error" and warning" when running but those can easily be addressed.
This is what my `.eslintrc.json` looks like

```json
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended"
  ],
  "env": {
    "node": true,
    "es6": true
  },
  "parserOptions": {
    "ecmaVersion": 2021
  },
  "rules": {
    "arrow-spacing": ["warn", { "before": true, "after": true }],
    "brace-style": ["error", "stroustrup", { "allowSingleLine": true }],
    "comma-dangle": ["error", "always-multiline"],
    "comma-spacing": "error",
    "comma-style": "error",
    "curly": ["error", "multi-line", "consistent"],
    "dot-location": ["error", "property"],
    "handle-callback-err": "off",
    "indent": "off",
    "keyword-spacing": "error",
    "max-nested-callbacks": ["error", { "max": 4 }],
    "max-statements-per-line": ["error", { "max": 2 }],
    "no-console": "off",
    "no-empty-function": "error",
    "no-floating-decimal": "error",
    "no-inline-comments": "error",
    "no-lonely-if": "off",
    "no-multi-spaces": "error",
    "no-multiple-empty-lines": [
      "error",
      { "max": 2, "maxEOF": 1, "maxBOF": 0 }
    ],
    "no-shadow": ["error", { "allow": ["err", "resolve", "reject"] }],
    "no-trailing-spaces": ["error"],
    "no-var": "error",
    "object-curly-spacing": ["error", "always"],
    "prefer-const": "error",
    "quotes": ["error", "single"],
    "semi": ["error", "always"],
    "space-before-blocks": "error",
    "space-before-function-paren": [
      "error",
      {
        "anonymous": "never",
        "named": "never",
        "asyncArrow": "always"
      }
    ],
    "space-in-parens": "error",
    "space-infix-ops": "error",
    "space-unary-ops": "error",
    "spaced-comment": "error",
    "yoda": "error"
  }
}
```

stinkyfi committed:
Linter is now running, removing info commands for now, and adding missing source to env variable

stinkyfi commented:
@AmadeusMoon I have a local branch with some changes, I cannot contribute to your fork. I also cannot fork your fork lol, because my code is a fork that already exist. I am dropping code changes in comments, or if you grant me access, I can make a new branch and create a pull request into your pull request.

coding inception going on here

AmadeusSlabu commented:
> @AmadeusMoon I have a local branch with some changes, I cannot contribute to your fork. I also cannot fork your fork lol, because my code is a fork that already exist. I am dropping code changes in comments, or if you grant me access, I can make a new branch and create a pull request into your pull request.
> 
> coding inception going on here

First of all just add me on discord if you have one AmadeusMoon , as for the code thing we have to discuss what you're planning to do and how, I have seen your changes, you want to add a crimes function and something about a waifu. Yes I stalked ur code it is my duty.

stinkyfi committed:
Adding space after //, per lint

stinkyfi committed:
fixing linter config

stinkyfi committed:
no more //

stinkyfi committed:
Merge branch 'main' of https://github.com/AmadeusMoon/MaddenBot-Typescript into stinky-dev

AmadeusSlabu committed:
Update open-ticket.ts to work

imported Client for type and added _client: Client to callback

AmadeusSlabu committed:
Update source.ts to work

Imported Client for type and added _client to callback

AmadeusSlabu committed:
Changes to source and ticket

AmadeusSlabu committed:
Merge branch 'main' of https://github.com/AmadeusMoon/MaddenBot-Typescript into main

AmadeusSlabu committed:
Changes to source

stinkyfi committed:
Fixing conflict and linter

stinkyfi committed:
fixing more conflicts

AmadeusSlabu committed:
Clean little import

AmadeusSlabu committed:
Changed most requires for imports
Around 53 problems left 3 errors and 50 warnings
Trying to figure the 3 errors caused by inline requires

stinkyfi committed:
Removing pipeline, need to revisit linting solution in future
